### DISCUSSION TITLE
Question: Handle Leadership Change in PullConsumers

### DISCUSSION THREAD

**Discussion by ludusrusso** [nats-io/nats.go#965] [Q&A]
Hi guys, I'm running a nats jetstream cluster on kubernetes with the official helm chart and 3 replicas. 
In some cases I notice an error of `Leadership Change` in pull consumer, here is an example code:

```golang
	ctx, cancel := context.WithTimeout(mainCtx, 10*time.Second)
	defer cancel()
	msgs, err := sub.Fetch(3, nats.Context(ctx))
	if err != nil {
		if errors.Is(err, nats.ErrTimeout) || errors.Is(err, context.DeadlineExceeded) {
			return nil
		}
		return fmt.Errorf("cannot fetch: %w", err)
	}
	
	// handle msgs
```

Hi was wondering, which is the best way to handle this errors? Should I recreate the subscriber or I can just retry to Fetch?

**Comment by wallyqs** [comment]
Short answer I think would be to retry, but also 'Leadership Change' is a [409 status](https://github.com/nats-io/nats-server/blob/d520a27c36f51ede2e52ded8956562d5f894c8bc/server/consumer.go#L1764) sent from the server and a temporary error that I think the client ought to handle internally similar to how 408 is being handled until the deadline expires:

https://github.com/nats-io/nats.go/blob/1714729895dfe3f5aa0cfee3984b89bebcfce23c/nats.go#L3319-L3320

https://github.com/nats-io/nats.go/blob/1714729895dfe3f5aa0cfee3984b89bebcfce23c/js.go#L2450-L2455

**Reply by ludusrusso** [reply]
Ok, so for now I can workaround it with this

```golang
if err != nil {
		if errors.Is(err, nats.ErrTimeout) || errors.Is(err, context.DeadlineExceeded) {
			return nil
		} else if err.Error() == "nats: Leadership Change" {
			logrus.Errorf("nats leadership change, restarting loop: %v", err)
			return nil
		}
		return fmt.Errorf("cannot fetch: %w", err)
	}
```

Maybe I can push a PR to handle this bug
